### ISSUE TITLE
Criterion benchmarks: ferrolearn vs sklearn performance comparison

### ISSUE THREAD

**Issue by dollspace-gay** [dollspace-gay/ferrolearn#14]
## Summary

Add criterion benchmarks for all ferrolearn algorithms across small/medium/large datasets in both f32 and f64, plus a Python comparison script that runs the same workloads through sklearn to produce a head-to-head performance comparison.

## Context

The design doc (REQ-28/29) requires criterion benchmarks but none exist. We have no data on whether ferrolearn is faster or slower than sklearn. The hypothesis is Rust wins on small data and tree methods, sklearn+MKL wins on large linear algebra — but that's a guess without numbers.

### Dataset sizes
| Label | Samples | Features |
|---|---|---|
| tiny | 50 | 5 |
| small | 1,000 | 10 |
| medium | 10,000 | 100 |
| large | 100,000 | 1,000 |

### Algorithms to benchmark

**Regressors** (fit + predict):
- LinearRegression
- Ridge
- Lasso (max_iter=1000)
- ElasticNet (max_iter=1000)

**Classifiers** (fit + predict):
- LogisticRegression (max_iter=1000)
- DecisionTreeClassifier
- RandomForestClassifier (n_estimators=100)
- KNeighborsClassifier (n_neighbors=5)
- GaussianNB

**Transformers** (fit + transform):
- StandardScaler
- PCA (n_components=10)

**Clusterers** (fit + predict):
- KMeans (n_clusters=8, n_init=10)

**Metrics** (compute only):
- accuracy_score, f1_score, r2_score, mean_squared_error, silhouette_score

### What to measure

For each (algorithm, dataset_size, dtype) triple:
1. **fit** time (median of N iterations, criterion handles this)
2. **predict/transform** time on test set (same size as train)
3. **fit+predict** combined (the realistic workflow)

### Rust side: criterion benches

```
ferrolearn-bench/
├── Cargo.toml          # depends on criterion + all ferrolearn crates
├── benches/
│   ├── regressors.rs
│   ├── classifiers.rs
│   ├── transformers.rs
│   ├── clusterers.rs
│   └── metrics.rs
└── src/
    └── lib.rs          # shared data generation helpers
```

Each bench uses `criterion::BenchmarkGroup` with the dataset size as parameter:

```rust
fn bench_linear_regression(c: &mut Criterion) {
    let mut group = c.benchmark_group("LinearRegression/fit");
    for (label, n, p) in [("tiny", 50, 5), ("small", 1000, 10), ("medium", 10000, 100)] {
        let (x, y) = generate_regression_data(n, p, 42);
        group.bench_with_input(BenchmarkId::new("f64", label), &(&x, &y), |b, (x, y)| {
            b.iter(|| LinearRegression::<f64>::new().fit(x, y).unwrap())
        });
    }
    group.finish();
}
```

### Python side: sklearn comparison

`scripts/benchmark_sklearn.py` — runs the same workloads with `timeit`:

```python
import timeit
from sklearn.linear_model import LinearRegression
from sklearn.datasets import make_regression

for n, p in [(50, 5), (1000, 10), (10000, 100), (100000, 1000)]:
    X, y = make_regression(n_samples=n, n_features=p, random_state=42)
    t = timeit.timeit(lambda: LinearRegression().fit(X, y), number=100)
    print(f"LinearRegression fit {n}x{p}: {t/100*1000:.2f} ms")
```

Output as JSON for automated comparison.

### Comparison report

A script that reads criterion JSON output + sklearn JSON output and produces:

```
Performance Comparison: ferrolearn vs sklearn
=============================================
                          tiny (50x5)    small (1Kx10)   medium (10Kx100)  large (100Kx1K)
LinearRegression fit      0.8x faster    1.2x faster     0.9x slower       1.1x slower
Ridge fit                 ...
DecisionTree fit          ...
RandomForest fit          ...
KMeans fit                ...
```

## Acceptance criteria
- [ ] Criterion benchmarks for all 12 algorithms at 3+ dataset sizes
- [ ] f64 benchmarks for all; f32 benchmarks for a subset
- [ ] Python sklearn comparison script with matching workloads
- [ ] Comparison report script that produces a readable table
- [ ] Results documented (even if sklearn wins on some workloads — honesty matters)
- [ ] `cargo bench` runs without errors
- [ ] Bench crate does not affect `cargo test --workspace`
